### ISSUE TITLE
Potential storage collision with STATE slot

### ISSUE THREAD

**Issue by abacabadabacaba** [near/near-sdk-rs#267]
In near-sdk-rs, there is a feature that automatically loads the contract state before a function is called and saves it afterwards. For this purpose, the storage slot with the key `STATE` is used. However, this is not documented, so the user may unintentionally use the same slot for something else, leading to incorrect behavior.

The usage can be either explicit (such as `env::storage_write(&b'STATE', something)`) or implicit through collections (such as `collections::LookupMap::<u32, Something>::new(vec![83]).insert(1163149652, something)`).

To fix this, the first step is to document the storage usage of this feature. The other thing that I would recommend to do is to change the storage key to an empty byte sequence. This has an advantage that it is guaranteed to not collide with any storage keys used by collections with a nonempty prefix.

**Comment by evgenykuzyakov** [comment]
Good feedback.

I think we should expose it using `#[near_bindgen(state_key=b"")`.
Update all examples to use `#[near_bindgen(state_key=b"")` and keep default `state_key` to `b"STATE"`.

**Comment by encody** [comment]
Has the `#[near_bindgen(state_key=b"")]` feature been implemented? I noticed that the default `"STATE"` key has been documented[^1][^2] (though I had to go digging to find those mentions). Since the concerns in the original issue haven't really been addressed (there's nothing preventing an uninformed author from accidentally `storage_write`ing over top of `"STATE"`), I wonder if it wouldn't be prudent to have a "safe" version of `storage_write` that checks that the key is not equal to `"STATE"` and a `storage_write_unchecked` (or something similar) that doesn't perform that check?

[^1]: https://www.near-sdk.io/contract-structure/nesting#generating-unique-prefixes-for-persistent-collections
[^2]: https://www.near-sdk.io/contract-interface/contract-mutability#mutable-functions

**Comment by austinabell** [comment]
Yes, this is definitely possible. It's a bit error-prone because it requires someone to know that they need to add this annotation to every `near_bindgen` attribute and not just the state struct which might be intuitive. Can explore other API types for this, but will add this to backlog.
